cmd/kube-proxy/app/server.go

```diff
@@ -689,8 +694,38 @@ func newProxyServer(logger klog.Logger, config *kubeproxyconfig.KubeProxyConfigu
 	return s, nil
 }
 
-// checkIPConfig confirms that s has proper configuration for its primary IP family.
-func checkIPConfig(s *ProxyServer, dualStackSupported bool) (error, bool) {
+// checkBadConfig checks for bad/deprecated configuation
+func checkBadConfig(s *ProxyServer) error {
+	var errors []error
+
+	clusterCIDRs := strings.Split(s.Config.ClusterCIDR, ",")
+	anyDualStackConfig := false
+	for _, config := range [][]string{clusterCIDRs, s.Config.NodePortAddresses, s.Config.IPVS.ExcludeCIDRs} {
+		if dual, _ := netutils.IsDualStackCIDRStrings(config); dual {
+			anyDualStackConfig = true
+			break
+		}
+	}
+
+	if !s.Config.NodePortAddressesPrimary {
+		cidrsByFamily := proxyutil.MapCIDRsByIPFamily(s.Config.NodePortAddresses)
+		if len(s.Config.NodePortAddresses) == 0 {
+			errors = append(errors, fmt.Errorf("nodePortAddresses and nodePortAddressesPrimary are both unset; NodePort connections will be accepted on all local IPs. Consider using `--nodeport-addresses-primary`"))
+		} else if anyDualStackConfig && len(cidrsByFamily[s.PrimaryIPFamily]) == len(s.Config.NodePortAddresses) {
+			errors = append(errors, fmt.Errorf("cluster appears to be dual-stack but nodePortAddresses contains only %s addresses; NodePort connections will be accepted on all local %s IPs", s.PrimaryIPFamily, proxyutil.OtherIPFamily(s.PrimaryIPFamily)))
+		}
+		// checkBadIPConfig() will warn if NodePortAddresses only has IPs of the
+		// non-primary family.
+	}
+
+	return utilerrors.NewAggregate(errors)
+}
+
+// checkBadIPConfig checks for bad configuration relative to s.PrimaryIPFamily.
+// Historically, we did not check most of the config options, so we cannot retroactively
+// make IP family mismatches in those options be fatal. When we add new options to check
+// here, we should make problems with those options be fatal.
+func checkBadIPConfig(s *ProxyServer, dualStackSupported bool) (err error, fatal bool) {
```

Comment:
Why isn't this called from `checkBadConfig` - the split seems artificial?

Reply:
I just tried merging them, and it makes the unit tests messy because the `checkBadIPConfig` results depend on `dualStackSupported`, but the `checkBadConfig` results don't...